### ISSUE TITLE
Incorrect "IDE0004 Cast is redundant" in ternary involving nullable and Task

### ISSUE THREAD

**Issue by drewnoakes** [dotnet/roslyn#60720]
**Version Used**: 

int.main

**Steps to Reproduce**:

```c#
#nullable enable

using System;
using System.Threading.Tasks;

Task<string?> M(string? s, Func<string, Task<string>> func)
{
    return string.IsNullOrWhiteSpace(s)
        ? Task.FromResult<string?>(null)
        : (Task<string?>)func(s)!;
    //    ^^^^^^^^^^^^^^^ IDE0004 Cast is redundant
}
```

https://sharplab.io/#v2:D4YgdgrgNlCGBGUCmACJYHILAChcAEAmARgIAYV9iBWAblwIDYAeKsgfgD4UBZACjbsUAZwA0laq2Jlx+Fm07cAZhDABjAJS4A3rhT7KAdkrSAdAElhAOWhQA8gCcA6gAsAlgBckAZQAOsNSQ+YS0cA3CUITlTADEHAHsAWwAlJGFoDykOTj5IGFCIgwAuFAF5aS4NFXVgjQBCejCDAHpm8IA9Tq7unpRzABEAUTIRgBYUAGFYYQ8UN2EUByQAE1Vl2DAPXABfIA

**Expected Behavior**:

Either
- invoking code fix produces correct code, or
- code is not flagged as redundant

**Actual Behavior**:

Code is marked as redundant, and removing it produces diagnostic:

> CS8619: Nullability of reference types in value of type 'Task<string>' doesn't match target type 'Task<string?>'.

**Comment by drocha87** [comment]
Also if you have code like this 

```csharp
```

**Comment by drocha87** [comment]
Also `IDE0004` is misleading in codes like this:

```csharp
public static float Example()
{
    var foo = 1;
    var bar = 2;
    return (float)foo / (float)bar; // IDE0004: suggests that cast is redundant
}
```
But you can clearly see that the result will be completely wrong if you remove the cast, since you are expecting the result to be `0.5` but without the cast it's `0`.

**Comment by CyrusNajmabadi** [comment]
@drocha87 either of those casts are redundant and can be removed. The feature supports this properly (including only removing one if you do 'fix all').

**Comment by drocha87** [comment]
I understand that the `fix all` fixes the code properly. But I could provide a little bit of more context. For example in vscode the above code will be like this:

![image](https://user-images.githubusercontent.com/31552590/164741585-e500cadd-5bf2-42ce-b77a-f5d8563ef3ec.png)

As you can see both `(float)` is "grey (faded)" which gives me the impression that both of them could be removed. 

Even roslyn is reporting them as two individual diagnostics which seems to me that cast redudancy is not expression-wise. 

That's why I'm saying it's misleading, because if you follow your instinct `Grey stuff can be removed` and remove both of them your code will subtly broken.
